_posts/2020-12-01-the-road-to-the-jit.md

```diff
@@ -0,0 +1,372 @@
+  the Erlang compiler that runs many passes.  Typically, when some of the code
+  for one pass had been converted to native code, the compiler was already running
+  the next pass.
+
+The later projects mitigated some of the issues in the previous projects.  For example,
+the compilation was reduced by doing more optimizations before invoking LLVM.
+Ultimately, though, it was decided to terminate the third and final tracing JIT project
+at the end of 2019.
+
+### The new JIT (also known as BeamAsm)
+
+After the end of the third tracing JIT project, [Lukas
+Larsson][lukas], having been involved in the last two tracing JIT
+projects, could not stop thinking about different approaches that
+might lead to a useful JIT.  The things that slowed down the previous
+approaches were the tracing to find hot code and the generation of
+optimized native code using LLVM.  Would it be possible to have a
+simpler JIT that didn't do tracing and did no or little optimization?
+
+In January 2020, salvaging some code from the third tracing JIT
+project, Lukas quickly built a prototype BEAM system that translated
+each BEAM instruction at load time to native code.  The resulting code
+was less optimized than LLVM-generated code because it would still use
+BEAM's stack and X registers (stored in memory), but the overhead for
+[instruction unpacking and instruction dispatch][overhead] was
+eliminated.
+
+[overhead]: http://blog.erlang.org/a-closer-look-at-the-interpreter/
+
+The initial benchmarks results were promising: about twice as fast
+compared to interpreted BEAM code, so Lukas extended the prototype
+so that it could handle more kind of BEAM instructions.
+
+[John Högberg][john] quickly became interested in the project and
+started to act as a sounding board.  Some time later, probably in
+March, John suggested that the new JIT should translate **all** loaded
+code to native code.  That way, there would be no need to support
+context switching between the BEAM interpreter and native code, which
+would make the design simpler and eliminate the cost for context
+switches.
+
+That was a gamble, of course. After all, it could turn out that the
+native code could be too large to be practically useful or decrease
+performance because it fitted badly in the code cache.  They decided
+that it was worth taking the risk and that it would probably be
+possible to optimize the size of the code later.  (*Spoiler*: At the time
+of writing, the native code generated by the JIT is about 10 percent
+larger than interpreted BEAM code.)
+
+Another change to the design was the tooling for generated the native
+code. In Lukas's prototype, the native code template for each
+instruction was contained in text files similar to the other files
+used by the loader.  That was inflexible, so it was decided to use
+some library that could generate native code.  While some pure C
+libraries could have been used, the C++ library AsmJIT was more
```

Comment:
Licensing was also played a major part in choosing which assembler to use. Some of the available assemblers have a copyleft license which does not work for us.